File: frontend/src/lib/api/vault-migration.ts
import { getJson, sendJson } from "./request";
import type { StorageProviderConfigValues } from "@/types";

export type VaultMigrationState =
  | "planned"
  | "copying"
  | "ready"
  | "activating"
  | "active"
  | "cleaned"
  | "discarded"
  | "paused"
  | "cutover_pending"
  | "draining"
  | "delta_copy"
  | "verifying"
  | "recovery_required"
  | "failed"
  | "complete";
export interface VaultMigrationPolicy {
  retention_days: number;
  concurrency: number;
  bandwidth_bytes_per_second: number | null;
}
export interface VaultMigrationAudit {
  id: number;
  state: string;
  critical_count: number;
  warning_count: number;
}
export interface VaultMigrationFailure {
  object_id: number | null;
  code: string;
  retryable: boolean;
}

export interface VaultMigrationRun {
  id: string;
  state: VaultMigrationState;
  plan_digest: string;
  backup_summary: { backup_id: string; source_ref: string | null; verified_at: string };
  objects: number;
  verified_objects: number;
  bytes: number;
  error_code: string | null;
  cleanup_after: string | null;
  cleanup_findings: { object_id: number | null; code: string }[];
  source_retained: boolean;
  expires_at: string;
  source: StorageProviderConfigValues;
  destination: StorageProviderConfigValues;
  capacity_warnings: string[];
  recovery_required: boolean;
  source_provider_ref: string;
  destination_provider_ref: string;
  capacity_resources: { role: string; required_bytes: number; available_bytes: number | null }[];
  policy: VaultMigrationPolicy;
  copied_objects: number;
  copied_bytes: number;
  verified_bytes: number;
  skipped_objects: number;
  skipped_bytes: number;
  failed_objects: number;
  failed_bytes: number;
  delta_objects: number;
  throughput_bytes_per_second: number | null;
  last_activity_at: string | null;
  retryable: boolean;
  phase_history: { phase: string; at: string }[];
  pre_audit: VaultMigrationAudit | null;
  post_audit: VaultMigrationAudit | null;
  full_audit: VaultMigrationAudit | null;
  cleanup_outcome: "retained_indefinitely" | "manual_cleanup" | "cleaned" | null;
  notification_events: number;
}
export interface VaultMigrationPreflight {
  destination: StorageProviderConfigValues;
  backup_id: string;
  backup_source_ref?: string | null;
  policy?: VaultMigrationPolicy;
}
const base = "/api/v1/storage/migrations";
const path = (id: string) => `${base}/${encodeURIComponent(id)}`;
export const listVaultMigrations = () => getJson<VaultMigrationRun[]>(base, { fresh: true });
export const getVaultMigration = (id: string) =>
  getJson<VaultMigrationRun>(path(id), { fresh: true });
export const preflightVaultMigration = (body: VaultMigrationPreflight) =>
  sendJson<VaultMigrationRun>(`${base}/preflight`, "POST", body);
export const startVaultMigration = (id: string, planDigest: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/start`, "POST", { plan_digest: planDigest });
export const advanceVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/advance`, "POST", {});
export const cutoverVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/cutover`, "POST", {});
export const recoverVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/recover`, "POST", {});
export const cleanupVaultMigration = (
  id: string,
  source: boolean,
  backup?: Pick<VaultMigrationPreflight, "backup_id" | "backup_source_ref">,
) =>
  sendJson<VaultMigrationRun>(`${path(id)}/cleanup`, "POST", {
    confirmation: id,
    source,
    ...backup,
  });

export const pauseVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/pause`, "POST", {});
export const resumeVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/resume`, "POST", {});
export const retryVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/retry`, "POST", {});
export const auditVaultMigration = (id: string) =>
  sendJson<VaultMigrationRun>(`${path(id)}/full-audit`, "POST", {});
export const retainVaultMigration = (id: string, removeCredentials = false) =>
  sendJson<VaultMigrationRun>(`${path(id)}/retain`, "POST", {
    remove_credentials: removeCredentials,
  });
export interface VaultMigrationReport extends VaultMigrationRun {
  resource_kind_totals: { resource_type: string; objects: number; bytes: number }[];
  recent_failures: VaultMigrationFailure[];
}
export const getVaultMigrationReport = (id: string) =>
  getJson<VaultMigrationReport>(`${path(id)}/report`, { fresh: true });
export async function downloadVaultMigrationReport(id: string): Promise<void> {
  const report = await getVaultMigrationReport(id);
  const url = URL.createObjectURL(
    new Blob([JSON.stringify(report, null, 2)], { type: "application/json" }),
  );
  const link = document.createElement("a");
  link.href = url;
  link.download = `printstash-migration-${encodeURIComponent(id)}.json`;
  document.body.appendChild(link);
  try {
    link.click();
  } finally {
    link.remove();
    URL.revokeObjectURL(url);
  }
}

File: backend/app/schemas/vault_migration.py
"""Safe public state for administrative Vault migration controls."""

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class MigrationPolicy(BaseModel):
    retention_days: int = Field(default=7, ge=0, le=3650)
    concurrency: int = Field(default=1, ge=1, le=4)
    bandwidth_bytes_per_second: int | None = Field(default=None, ge=1024)


class MigrationCleanupFinding(BaseModel):
    object_id: int | None
    code: str


class MigrationBackupSummary(BaseModel):
    backup_id: str
    source_ref: str | None
    archive_sha256: str
    vault_target_ref: str
    created_at: datetime
    verified_at: datetime


class MigrationRunRead(BaseModel):
    id: str
    state: Literal[
        "planned",
        "copying",
        "ready",
        "paused",
        "cutover_pending",
        "draining",
        "delta_copy",
        "verifying",
        "activating",
        "active",
        "cleaned",
        "discarded",
        "failed",
        "recovery_required",
        "complete",
    ]
    plan_digest: str
    backup_summary: MigrationBackupSummary
    source: dict[str, object]
    destination: dict[str, object]
    source_provider_ref: str
    destination_provider_ref: str
    objects: int = Field(ge=0)
    verified_objects: int = Field(ge=0)
    bytes: int = Field(ge=0)
    copied_objects: int
    copied_bytes: int
    verified_bytes: int
    skipped_objects: int
    skipped_bytes: int
    failed_objects: int
    failed_bytes: int
    delta_objects: int
    throughput_bytes_per_second: float
    last_activity_at: datetime
    retryable: bool
    policy: MigrationPolicy
    phase_history: list[dict[str, str]]
    pre_audit: dict[str, object] | None
    post_audit: dict[str, object] | None
    full_audit: dict[str, object] | None
    notification_events: int
    error_code: str | None
    expires_at: datetime
    cleanup_after: datetime | None
    cleanup_findings: list[MigrationCleanupFinding]
    cleanup_outcome: str | None
    source_retained: bool
    recovery_required: bool
    capacity_warnings: list[str]
    capacity_resources: list[dict[str, object]]
    resource_kind_totals: list[dict[str, object]]
    recent_failures: list[dict[str, object]]
